File: 018removeNode.js
const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

const removeNode = (head, targetVal) => {
  // todo
  let current = head
  let prev = head
  let moded = false

  if (head.val === targetVal){
    let newHead = head.next
    head = null
    return newHead
  }

  while (current !== null && moded === false) {
    // console.log(current.next)
    // console.log(targetVal)
    if (current.val === targetVal){

      prev.next = current.next
      moded = true

    } else {
      prev = current

    }
      current = current.next

  }
  return head
};

removeNode(a, "c");
// a -> b -> d -> e -> f

File: 018remove_node.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.next = None

# iterative solution
def remove_node(head, target_val):
  current = head
  prev = None

  if current.val == target_val:
    return head.next

  while current is not None:
    if current.val == target_val:
      prev.next = current.next
      break

    prev = current
    current = current.next

  return head

# n = number of nodes
# Time: O(n)
# Space: O(1)

# # recursive solution
# def remove_node(head, target_val):
#   if head is None:
#     return None

#   if head.val == target_val:
#     return head.next

#   head.next = remove_node(head.next, target_val)
#   return head

# n = number of nodes
# Time: O(n)
# Space: O(n)
